File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.java
class Solution {
    public int shipWithinDays(int[] weights, int D) {
        int left = 0, right = 0;
        for (int w = 0; w < weights.length; w++) {
            left = Math.max(left, weights[w]);
            right += weights[w];
        }
        while (left < right) {
            int mid = (left + right) / 2, need = 1, cur = 0;
            for (int w = 0; w < weights.length; w++) {
                if (cur + weights[w] > mid) {
                    need += 1;
                    cur = 0;
                }
                cur += weights[w];
            }
            if (need > D) left = mid + 1;
            else right = mid;
        }
        return left;
    }
}

File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.py
class Solution:
    def shipWithinDays(self, piles: List[int], h: int) -> int:

        start = 0
        end = start
        for num in piles:
            start = max(start,num)
            end += num

        ans = end
        while start <= end:
            mid = (start + end) // 2
            count = 1
            bun = mid
            i = 0
            su = 0
            while i < len(piles):
                
                if piles[i] + su > bun:
                    su = 0
                    count += 1
                su += piles[i]
                i += 1
            if count > h:
                start = mid + 1
            else:
                end = mid - 1
                ans = min(ans, mid)
        
        return ans
                